ql/src/test/org/apache/hadoop/hive/ql/materializedview/TestHeartbeatTxnRangeFunction.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.materializedview;
+
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.api.AbortTxnRequest;
+import org.apache.hadoop.hive.metastore.api.HeartbeatTxnRangeRequest;
+import org.apache.hadoop.hive.metastore.api.HeartbeatTxnRangeResponse;
+import org.apache.hadoop.hive.metastore.api.MetaException;
+import org.apache.hadoop.hive.metastore.api.NoSuchTxnException;
+import org.apache.hadoop.hive.metastore.api.OpenTxnRequest;
+import org.apache.hadoop.hive.metastore.api.TxnAbortedException;
+import org.apache.hadoop.hive.metastore.txn.TxnHandler;
+import org.apache.hadoop.hive.metastore.txn.TxnStore;
+import org.apache.hadoop.hive.metastore.txn.TxnUtils;
+import org.apache.hadoop.hive.metastore.utils.TestTxnDbUtil;
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.core.LoggerContext;
+import org.apache.logging.log4j.core.config.Configuration;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotEquals;
+import static org.junit.Assert.assertNotNull;
+
+public class TestHeartbeatTxnRangeFunction {
+  static final private String CLASS_NAME = TxnHandler.class.getName();
+
+  private final HiveConf conf = new HiveConf();
+  private TxnStore txnHandler;
+
+  public TestHeartbeatTxnRangeFunction() throws Exception {
+    TestTxnDbUtil.setConfValues(conf);
+    TestTxnDbUtil.prepDb(conf);
+    LoggerContext ctx = (LoggerContext) LogManager.getContext(false);
+    Configuration conf = ctx.getConfiguration();
+    conf.getLoggerConfig(CLASS_NAME).setLevel(Level.DEBUG);
+    ctx.updateLoggers(conf);
+    tearDown();
+  }
+
+  @Before
+  public void setUp() throws Exception {
+    txnHandler = TxnUtils.getTxnStore(conf);
```

Comment:
why do you need a new instance of txnHandler for every test?

Reply:
It is a good question. I got the inspiration from TestTxnHandler. TBH, I had the exact same question. But decided to keep it here assuming there is something that I should know about.

See: https://github.com/apache/hive/blob/6b48995442a514cd946663b2c0343e772430761e/ql/src/test/org/apache/hadoop/hive/metastore/txn/TestTxnHandler.java#L1860

Reply:
I checked it. Our tests can run with static txnHandler so I refactored it. 
TBH, my first thought was that it can lead to flaky test. Usually, sharing any kind of state between tests is a bad pattern as test won't be isolated.

Reply:
txnHandler is just a client here